support using fa4 on deepseek on blackwell

cicirori opened the pull request:
This is an initial PR for FA4 backend.

1. Get FA4 from the original repository into the sgl kernel and share the interface with FA3, controlled by the `ver` parameter.
2. Add support for using FA4 as the prefill backend with dpsk-like model + Blackwell.

Based on the current interface compatibility of Flash Attention 4,I chose to share the implementation with the FA3 attention backend.

cicirori committed:
add experimental fa4 support for dpsk-like model

cicirori committed:
refine

cicirori committed:
fix windows size && add some assert

cicirori committed:
add tests

cicirori commented:
test gpu: B200

launch command
```
python3 -m sglang.launch_server   --model-path deepseek-v3-FP4   --trust-remote-code   --quantization modelopt_fp4   --tp-size 4   --ep-size 1     --model-loader-extra-config '{"enable_multithread_load": true, "num_threads": 8}'    
# w/ --prefill-attention-backend fa4 if enable fa4
```
benchmark command
```
python3 -m sglang.bench_serving --backend sglang --dataset-name random --num-prompts 256 --random-input 2048 --random-output 1 --random-range-ratio 0.5 --flush-cache --max-concurrency 1
```


```
FA4
---------------Time to First Token----------------
Mean TTFT (ms):                          117.73
Median TTFT (ms):                        119.51
P99 TTFT (ms):                           142.93

Flashinfer
---------------Time to First Token----------------
Mean TTFT (ms):                          131.84
Median TTFT (ms):                        134.83
P99 TTFT (ms):                           169.19
```

cicirori committed:
fix

cicirori committed:
fix ci

cicirori committed:
comments

zhyncs commented on python/sglang/srt/models/deepseek_v2.py:
```diff
@@ -1732,7 +1735,10 @@ def forward_normal_chunked_kv_prepare(
 
     def forward_normal_chunked_kv_core(self, q, k, v, forward_batch):
-        has_extend_prefix = any(forward_batch.extend_prefix_lens_cpu)
+        has_extend_prefix = forward_batch.extend_prefix_lens_cpu and any(
```
```
[2025-09-02 21:34:26 DP1 TP2 EP2] TpModelWorkerClient hit an exception: Traceback (most recent call last):
  File "/public_sglang_ci/runner-l1a-gpu-4567/_work/sglang/sglang/python/sglang/srt/managers/tp_worker_overlap_thread.py", line 141, in forward_thread_func
    self.forward_thread_func_()
  File "/usr/local/lib/python3.10/dist-packages/torch/utils/_contextlib.py", line 120, in decorate_context
    return func(*args, **kwargs)
  File "/public_sglang_ci/runner-l1a-gpu-4567/_work/sglang/sglang/python/sglang/srt/managers/tp_worker_overlap_thread.py", line 176, in forward_thread_func_
    self.worker.forward_batch_generation(
  File "/public_sglang_ci/runner-l1a-gpu-4567/_work/sglang/sglang/python/sglang/srt/managers/tp_worker.py", line 239, in forward_batch_generation
    logits_output, can_run_cuda_graph = self.model_runner.forward(
  File "/public_sglang_ci/runner-l1a-gpu-4567/_work/sglang/sglang/python/sglang/srt/model_executor/model_runner.py", line 1763, in forward
    output = self._forward_raw(
  File "/public_sglang_ci/runner-l1a-gpu-4567/_work/sglang/sglang/python/sglang/srt/model_executor/model_runner.py", line 1808, in _forward_raw
    ret = self.forward_extend(
  File "/public_sglang_ci/runner-l1a-gpu-4567/_work/sglang/sglang/python/sglang/srt/model_executor/model_runner.py", line 1708, in forward_extend
    return self.model.forward(
  File "/usr/local/lib/python3.10/dist-packages/torch/utils/_contextlib.py", line 120, in decorate_context
    return func(*args, **kwargs)
  File "/public_sglang_ci/runner-l1a-gpu-4567/_work/sglang/sglang/python/sglang/srt/models/deepseek_v2.py", line 2220, in forward
    hidden_states = self.model(
  File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py", line 1773, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py", line 1784, in _call_impl
    return forward_call(*args, **kwargs)
  File "/public_sglang_ci/runner-l1a-gpu-4567/_work/sglang/sglang/python/sglang/srt/models/deepseek_v2.py", line 2089, in forward
    hidden_states, residual = layer(
  File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py", line 1773, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py", line 1784, in _call_impl
    return forward_call(*args, **kwargs)
  File "/public_sglang_ci/runner-l1a-gpu-4567/_work/sglang/sglang/python/sglang/srt/models/deepseek_v2.py", line 1879, in forward
    hidden_states = self.self_attn(
  File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py", line 1773, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py", line 1784, in _call_impl
    return forward_call(*args, **kwargs)
  File "/public_sglang_ci/runner-l1a-gpu-4567/_work/sglang/sglang/python/sglang/srt/models/deepseek_v2.py", line 1091, in forward
    return self.forward_core(s)
  File "/public_sglang_ci/runner-l1a-gpu-4567/_work/sglang/sglang/python/sglang/srt/models/deepseek_v2.py", line 1145, in forward_core
    return self.forward_normal_chunked_kv_core(*inner_state)
  File "/public_sglang_ci/runner-l1a-gpu-4567/_work/sglang/sglang/python/sglang/srt/models/deepseek_v2.py", line 1738, in forward_normal_chunked_kv_core
    has_extend_prefix = forward_batch.extend_prefix_lens_cpu and any(
RuntimeError: Boolean value of Tensor with more than one value is ambiguous
```

zhyncs replied:
ref https://github.com/sgl-project/sglang/blob/369b143366a4c58cc37f81bd4141c8794f26e56f/.github/workflows/pr-test.yml#L347-L365

cicirori committed:
for cr

zhyncs committed:
Merge branch 'main' into feat/fa4_for_dpsk

cicirori committed:
Merge branch 'main' into feat/fa4_for_dpsk

zhyncs committed:
upd

zhyncs committed:
Merge branch 'main' into feat/fa4_for_dpsk

zhyncs committed:
upd

zhyncs committed:
Merge branch 'main' into feat/fa4_for_dpsk

fzyzcjy committed:
Merge branch 'main' into feat/fa4_for_dpsk

zhyncs committed:
Merge branch 'main' into feat/fa4_for_dpsk

fzyzcjy committed:
Merge branch 'main' into feat/fa4_for_dpsk

cicirori committed:
Merge branch 'main' into feat/fa4_for_dpsk

fzyzcjy committed:
Merge branch 'main' into feat/fa4_for_dpsk

cicirori committed:
Merge remote-tracking branch 'origin/main' into feat/fa4_for_dpsk

cicirori committed:
Merge remote-tracking branch 'origin/feat/fa4_for_dpsk' into feat/fa4_for_dpsk

cicirori committed:
reduce cute dsl log

cicirori committed:
add warmup at first time

zhyncs committed:
Merge branch 'main' into feat/fa4_for_dpsk

cicirori commented:
I disabled the Cute DSL logs and added a more concise log for compile triggers.
On the first call, I also added some known recompile rules that can trigger recompilation.

Test results show no correctness issues and no explicit recompilation, but the second run is still slightly faster (I flushed the cache between runs).

- first run
```
limit: None, num_fewshot: 8, batch_size: 128
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     8|exact_match|↑  |0.9401|±  |0.0065|
|     |       |strict-match    |     8|exact_match|↑  |0.9318|±  |0.0069|


real    2m48.260s
user    0m18.324s
sys     0m1.515s
```

- second run
```
limit: None, num_fewshot: 8, batch_size: 128
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     8|exact_match|↑  |0.9439|±  |0.0063|
|     |       |strict-match    |     8|exact_match|↑  |0.9340|±  |0.0068|


real    2m18.535s
user    0m18.703s
sys     0m1.447s
```

It feels like the outer compile-generated kernel may also have some heuristic cache related to input shape. Since FA4 does not trigger recompilation when input length changes, if this is indeed an issue, we might need to investigate further.

zhyncs commented:
```
python3 -m sglang.launch_server --model-path nvidia/DeepSeek-V3-0324-FP4 --tp 4 --attention-backend trtllm_mla --moe-runner-backend flashinfer_trtllm --quantization modelopt_fp4 --speculative-algorithm EAGLE --speculative-num-steps 3 --speculative-eagle-topk 1 --speculative-num-draft-tokens 4 --prefill-attention-backend fa4
```

```
  File "/usr/local/lib/python3.12/dist-packages/torch/nn/modules/module.py", line 1784, in _call_impl                                                                             [220/564]
    return forward_call(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/sgl-workspace/sglang/python/sglang/srt/models/deepseek_v2.py", line 2369, in forward
    hidden_states, residual = layer(
                              ^^^^^^
  File "/usr/local/lib/python3.12/dist-packages/torch/nn/modules/module.py", line 1773, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/dist-packages/torch/nn/modules/module.py", line 1784, in _call_impl
    return forward_call(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/sgl-workspace/sglang/python/sglang/srt/models/deepseek_v2.py", line 2098, in forward
    hidden_states = self.self_attn(
                    ^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/dist-packages/torch/nn/modules/module.py", line 1773, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/dist-packages/torch/nn/modules/module.py", line 1784, in _call_impl
    return forward_call(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/sgl-workspace/sglang/python/sglang/srt/models/deepseek_v2.py", line 1208, in forward
    return self.forward_core(s)
           ^^^^^^^^^^^^^^^^^^^^
  File "/sgl-workspace/sglang/python/sglang/srt/models/deepseek_v2.py", line 1270, in forward_core
    return self.forward_normal_chunked_kv_core(*inner_state)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/sgl-workspace/sglang/python/sglang/srt/models/deepseek_v2.py", line 1946, in forward_normal_chunked_kv_core
    has_extend_prefix = any(forward_batch.extend_prefix_lens_cpu)
```

yuho8818 commented:
It seems that EP is not supported.
```
python3 -m sglang.launch_server \
  --model-path /disk1/modelscope/DeepSeek-V3-0324-FP4/ \
  --trust-remote-code \
  --quantization modelopt_fp4 \
  --tp-size 4 \
  --model-loader-extra-config '{"enable_multithread_load": true, "num_threads": 8}' \
  --prefill-attention-backend fa4 \
  --dp-size 4 \
  --enable-dp-attention \
  --enable-dp-lm-head \
  --ep-size 4
```


```
Multi-thread loading shards:  95% Completed | 76/80 [00:27<00:00,  4.92it/s]
Multi-thread loading shards:  96% Completed | 77/80 [00:27<00:00,  5.47it/s]
Multi-thread loading shards:  98% Completed | 78/80 [00:27<00:00,  4.95it/s]
Multi-thread loading shards:  99% Completed | 79/80 [00:28<00:00,  4.98it/s]
Multi-thread loading shards: 100% Completed | 80/80 [00:29<00:00,  2.33it/s]
Multi-thread loading shards: 100% Completed | 80/80 [00:29<00:00,  2.75it/s]

[2025-09-26 03:50:48 DP1 TP1 EP1] Scheduler hit an exception: Traceback (most recent call last):
  File "/sgl-workspace/sglang/python/sglang/srt/managers/scheduler.py", line 2816, in run_scheduler_process
    scheduler = Scheduler(
                ^^^^^^^^^^
  File "/sgl-workspace/sglang/python/sglang/srt/managers/scheduler.py", line 353, in __init__
    self.tp_worker = TpWorkerClass(
                     ^^^^^^^^^^^^^^
  File "/sgl-workspace/sglang/python/sglang/srt/managers/tp_worker_overlap_thread.py", line 66, in __init__
    self.worker = TpModelWorker(
                  ^^^^^^^^^^^^^^
  File "/sgl-workspace/sglang/python/sglang/srt/managers/tp_worker.py", line 97, in __init__
    self.model_runner = ModelRunner(
                        ^^^^^^^^^^^^
  File "/sgl-workspace/sglang/python/sglang/srt/model_executor/model_runner.py", line 288, in __init__
    self.initialize(min_per_gpu_memory)
  File "/sgl-workspace/sglang/python/sglang/srt/model_executor/model_runner.py", line 335, in initialize
    self.load_model()
  File "/sgl-workspace/sglang/python/sglang/srt/model_executor/model_runner.py", line 774, in load_model
    self.model = get_model(
                 ^^^^^^^^^^
  File "/sgl-workspace/sglang/python/sglang/srt/model_loader/__init__.py", line 28, in get_model
    return loader.load_model(
           ^^^^^^^^^^^^^^^^^^
  File "/sgl-workspace/sglang/python/sglang/srt/model_loader/loader.py", line 494, in load_model
    self.load_weights_and_postprocess(
  File "/sgl-workspace/sglang/python/sglang/srt/model_loader/loader.py", line 513, in load_weights_and_postprocess
    quant_method.process_weights_after_loading(module)
  File "/sgl-workspace/sglang/python/sglang/srt/layers/quantization/modelopt_quant.py", line 1243, in process_weights_after_loading
    (w13_input_scale * w13_weight_scale_2).to(torch.float32),
     ~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~
RuntimeError: The size of tensor a (256) must match the size of tensor b (64) at non-singleton dimension 0

Exception ignored in atexit callback: <function move_cutlass_compiled_cache at 0xfffb95f1afc0>
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/dist-packages/torch/_inductor/codegen/cuda/cutlass_utils.py", line 39, in move_cutlass_compiled_cache
    if not os.path.exists(cutlass.CACHE_FILE):
                          ^^^^^^^^^^^^^^^^^^
AttributeError: module 'cutlass' has no attribute 'CACHE_FILE'
[rank1]:[W926 03:50:49.778000071 ProcessGroupNCCL.cpp:1538] Warning: WARNING: destroy_process_group() was not called before program exit, which can leak resources. For more info, please see https://pytorch.org/docs/stable/distributed.html#shutdown (function operator())
```
